src/server/worker/pipeline/transform/worker.go

```diff
@@ -16,18 +18,251 @@ import (
+}
+
+func processComputeSerialDatumsTask(pachClient *client.APIClient, task *ComputeSerialDatumsTask) (*types.Any, error) {
+	dits := []datum.Iterator{
+		datum.NewCommitIterator(pachClient, task.BaseMetaCommit),
+		datum.NewFileSetIterator(pachClient, task.FileSetId),
+	}
+	var deleteFileSetID string
+	var skipped int64
+	outputFileSetID, err := withDatumFileSet(pachClient, func(outputSet *datum.Set) error {
+		var err error
+		deleteFileSetID, err = withDatumFileSet(pachClient, func(deleteSet *datum.Set) error {
+			return datum.Merge(dits, func(metas []*datum.Meta) error {
+				if len(metas) == 1 {
+					// Datum was processed in the parallel step.
+					if proto.Equal(metas[0].Job, task.Job) {
+						return nil
+					}
+					// Datum only exists in the parent job.
+					return deleteSet.UploadMeta(metas[0])
+				}
+				// Check if a skippable datum was successfully processed by the parent.
+				if !task.NoSkip && skippableDatum(metas[1], metas[0]) {
+					skipped++
+					return nil
+				}
+				if err := deleteSet.UploadMeta(metas[0]); err != nil {
+					return err
+				}
+				return outputSet.UploadMeta(metas[1], datum.WithPrefixIndex())
+			})
+		})
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+	return serializeComputeSerialDatumsTaskResult(&ComputeSerialDatumsTaskResult{
+		FileSetId:       outputFileSetID,
+		DeleteFileSetId: deleteFileSetID,
+		Skipped:         skipped,
+	})
+}
+
+func processCreateDatumSetsTask(driver driver.Driver, task *CreateDatumSetsTask) (*types.Any, error) {
+	setSpec, err := createSetSpec(driver, task.FileSetId)
+	if err != nil {
+		return nil, err
+	}
+	pachClient := driver.PachClient()
+	var inputFileSetsId string
+	resp, err := pachClient.WithCreateFileSetClient(func(mf client.ModifyFile) error {
+		return pachClient.WithRenewer(func(ctx context.Context, renewer *renew.StringSet) error {
+			pachClient := pachClient.WithCtx(ctx)
+			dit := datum.NewFileSetIterator(pachClient, task.FileSetId)
+			storageRoot := filepath.Join(driver.InputDir(), client.PPSScratchSpace, uuid.NewWithoutDashes())
+			var count int64
+			if err := datum.CreateSets(dit, storageRoot, setSpec, func(upload func(client.ModifyFile) error) error {
+				resp, err := pachClient.WithCreateFileSetClient(func(mf client.ModifyFile) error {
+					return upload(mf)
+				})
+				if err != nil {
+					return err
+				}
+				if err := renewer.Add(pachClient.Ctx(), resp.FileSetId); err != nil {
+					return err
+				}
+				name := fmt.Sprintf("%016d", count)
+				count++
+				input, err := serializeDatumSet(&DatumSet{
+					JobID:        task.Job.ID,
+					OutputCommit: task.OutputCommit,
+					FileSetId:    resp.FileSetId,
+				})
+				if err != nil {
+					return err
+				}
+				data, err := proto.Marshal(input)
+				if err != nil {
+					return errors.EnsureStack(err)
+				}
+				return errors.EnsureStack(mf.PutFile(name, bytes.NewReader(data)))
```

Comment:
so, if i'm understanding correctly, the change here is that the computed datum sets get written to/read from a fileset before the actual datum set tasks are created, which slows things down a little in order to add another checkpoint step

Reply:
That is true, but the performance hit is minimal considering it is one file per datum set. This model will enable more significant performance improvements down the line (i.e. parallel and distributed preprocessing).

Reply:
Agreed, I wasn't questioning the tradeoff, just making sure I wasn't missing anything